### DISCUSSION TITLE
thank you for writing pytest-console-scripts

### DISCUSSION THREAD

**Discussion by epogrebnyak** [kvas-it/pytest-console-scripts#81] [General]
I'm stuck with testing a click app with CliRunner in a case where a script executes when run from console, but fails as pytest with ambiguous message, so I was looking for a pytest plugin desperately. 

Very intuitive interface of  pytest-console-scripts and docs addressing the user needs,  `result.print()` should be a savior.

Thank you for putting this valuable project together.

**Comment by HexDecimal** [comment]
Discussions is probably better for this. Unless this wasn't just an appreciation post.

I'm glad pytest-console-scripts is working well for you.  I started using it myself since I was maintaining a project which had its own script runner implementation and I really wanted it to be in a central place rather than having this giant specialized tool just sitting there in the tests.  I was able covert everything to use pytest-console-scripts pretty quickly and easily.

Printing the stdout/stderr outputs so that they show up on pytest test failures is definitely the best feature of it.

**Comment by epogrebnyak** [comment]
Thanks for moving the post to discussions. Here is nice output I got:

```
script_runner = <ScriptRunner inprocess>, tmp_path = WindowsPath('C:/Users/epogr/AppData/Local/Temp/pytest-of-epogr/pytest-74/test_with_runner_inprocess_0')

    def test_with_runner(script_runner, tmp_path):
        write_toml(tmp_path)
        result = script_runner.run(['abacus', '--project-folder', str(tmp_path), "fire"])
        assert result.returncode == 0
>       assert result.stdout == "{'chart': 'chart.json', 'start': 'start_balances.json', 'entries': 'entries.json', 'rename': 'rename.json'}"
E       assert "{'chart': 'c...name.json'}\n" == "{'chart': 'c...rename.json'}"
E         - {'chart': 'chart.json', 'start': 'start_balances.json', 'entries': 'entries.json', 'rename': 'rename.json'}
E         + {'chart': 'chart.json', 'start': 'start_balances.json', 'entries': 'entries.json', 'rename': 'rename.json'}
E         ?                                                                                                            +

tests\test_cli.py:24: AssertionError
------------------------------------------------------------------- Captured stdout call -------------------------------------------------------------------- # Running console script: ['abacus', '--project-folder', 'C:\\Users\\epogr\\AppData\\Local\\Temp\\pytest-of-epogr\\pytest-74\\test_with_runner_inprocess_0', 'fire']
# Script return code: 0
# Script stdout:
{'chart': 'chart.json', 'start': 'start_balances.json', 'entries': 'entries.json', 'rename': 'rename.json'}

# Script stderr:
```

click.CliRunner was obfuscating improper formatting of a toml file as a file error without a chance of finding where the error was, I was able to fix with pytest-console-scripts in munites (couple hours spent with click.CliRunner on this). I also like the cleanliness of interface
`script_runner.run()`.

**Comment by epogrebnyak** [comment]
As a contribution idea I could do a PR with some notes to README.md about using `pytest-console-scripts` with poetry/pyproject.toml as not everthing is setup.py these days.

**Reply by HexDecimal** [reply]
I considered something like that but I wasn't sure how to structure it.

Every tool including setuptools is using `pyproject.toml` these days. So it'd make sense to show it in the readme, but I'm not sure people come to `pytest-console-scripts` to learn how to create console scripts, they come because they already have scripts and are having issues testing them.

**Reply by kvas-it** [reply]
We could consider replacing (or augmenting) the `setup.py` example in the "Usage" section of the `README` with an example using `pyproject.toml`. Putting things in `pyproject.toml` makes things more predictable (because then it's just data) and it's probably the best practice these days.

**Reply by epogrebnyak** [reply]
> ... but I'm not sure people come to `pytest-console-scripts` to learn how to create console scripts, they come because they already have scripts and are having issues testing them.

People often come confused and not fully sure which workflow to take, like myself, so extra information may help. There is so much confusion in the Python ecosystem about packaging, I think any sensible advice in README is useful. In specific I was thinking about pyproject.toml example for console scripts and poetry install / poetry run pytest workflow (I think you already describing it in readme ). 


Another idea is adding a section projects using pytest-console-scripts, which one can find in GitHub search.

**Reply by HexDecimal** [reply]
`pip install -e .` should work just as well with the latest Pip/Poetry. I agree that in general more documentation is better if it's segmented properly.

Showing projects using it might be a good idea, but I'm not sure what makes a good example project.

**Reply by epogrebnyak** [reply]
> `pip install -e .` should work just as well with the latest Pip/Poetry. I agree that in general more documentation is better if it's segmented properly.

I noticed a bit heated discussion about no `pip install -e .` feature for poetry with a summary [here](https://github.com/python-poetry/poetry/issues/34#issuecomment-1193365526). I should check where the latest `poetry` version stands on this. 

Will come back with my suggestions for README.md for pyproject.toml (mirror setup.py code) and poetry usage (how to ensure command line script is installed before test - in poetry that should be `poetry install` to install package and command line script and `poetry run pytest` to enter the environment).

**Reply by epogrebnyak** [reply]
> `pip install -e .` should work just as well with the latest Pip/Poetry. I agree that in general more documentation is better if it's segmented properly.

https://github.com/python-poetry/poetry/issues/34#issuecomment-1054626460 says: __poetry-core 1.0.8, I can now successfully run pip install -e . with the following block in my pyproject.toml:__ 

So yes, poetry package is installable with `pip install -e .`!

**Comment by kvas-it** [comment]
Thanks for giving this project a try, @epogrebnyak!

If you've found it recently, you got lucky -- pytest-console-scripts just has got a lot nicer with the many contributions from @HexDecimal :)

**Reply by epogrebnyak** [reply]
@HexDecimal, much appreciated! 

As a side note how serious is mentioning Python 4 in readme? What a reader should take from this?

**Reply by HexDecimal** [reply]
Do you mean pytest 4?  There isn't a Python 4.

**Reply by epogrebnyak** [reply]
> Thanks for giving this project a try, @epogrebnyak!

It was literally saving the cli interface for an app, I was at a verge of dumping click because I could not get tests to pass, found myself rescued! 

As sharing some dev experience: passing the env and cwd explicitly when invoking a script for testing is so clean, allows focus on  questions other than "What's cwd now? How should I set it? Will it hurt elsewhere if I set it, etc", thank for providing it this way!

**Reply by epogrebnyak** [reply]
> Do you mean pytest 4?  There isn't a Python 4.

My wrong! I was reading too fast.
